datameta/validation.py

```diff
@@ -0,0 +1,204 @@
+    val_errors += [
+        ({ 'site' : file_id, 'uuid' : file_id }, None, "Access denied")
+        for file_id, db_file in db_files.items()
+        if db_file is not None and not authz.submit_file(auth_user, db_file)
+    ]
+
+    # Collect missing metadatasets
+    val_errors += [
+        ({ 'site' : mset_id, 'uuid' : mset_id }, None, "Not found")
+        for mset_id, db_mset in db_msets.items()
+        if db_mset is None
+    ]
+    # Collect authorization issues
+    val_errors += [
+        ({ 'site' : mset_id, 'uuid' : mset_id }, None, "Access denied")
+        for mset_id, db_mset in db_msets.items()
+        if db_mset is not None and not authz.submit_mset(auth_user, db_mset) ]
+
+    # If we collected any val_errors so far, raise 400 to avoid exposing internals
+    # about data the user should not be able to access
+    if val_errors:
+        entities, fields, messages = zip(*val_errors)
+        raise errors.get_validation_error(messages=messages, fields=fields, entities=entities)
+
+def validate_submission_association(db_files, db_msets):
+    """Validates a submission with regard to
+
+    - All submitted files being associated to metadata
+    - All files referenced in the metadata being part of the submission
+    - All referenced entities have not been submitted before
+
+    Returns (tuple):
+        f_names_obj - A dict mapping file names to database file object
+        ref_fnames  - A dict mapping file names referenced by metadatumrecords to the metadatumrecord
+        errors      - A list of tuples (entity, field, message) describing the errors that occurred
+    """
+    errors = []
+    # Collect files with no data associated
+    errors += [ (db_file, None, "No data uploaded") for file_id, db_file in db_files.items() if db_file.content_uploaded==False ]
+    # Collect files which already have other metadata associated
+    errors += [ (db_file, None, "Already submitted") for file_id, db_file in db_files.items() if db_file.metadatumrecord is not None ]
+    # Collect metadatasets that were already submitted
+    errors += [ (db_mset, None, "Already submitted") for mset_id, db_mset in db_msets.items() if db_mset.submission_id is not None ]
+
+    # Collect the file names of the provided files
+    f_names_obj = defaultdict(list)
+    for db_file in db_files.values():
+        if db_file is not None:
+            f_names_obj[db_file.name].append(db_file)
+
+    # Make sure the filenames are unique
+    for fname, db_objs in f_names_obj.items():
+        if len(db_objs) > 1:
+            errors += [ (db_file, None, "Filename occurs multiple times among provided files") for db_file in db_objs ]
+
+    # Collect the file names referenced by the metadata sets - null values are
+    # not considered here, neither are records for which the file has already
+    # been linked (this case can occur when this function is called during a
+    # service execution)
+    mdat_fnames_obj = defaultdict(list)
```

Comment:
is there a specific reason why a `defaultdict` is used rather than `dict` with `setdefault`? (genuine interest)